salt/config/__init__.py

```diff
@@ -840,6 +840,13 @@ def _gather_buffer_space():
     # Always generate minion id in lowercase.
     'minion_id_lowercase': bool,
 
+    # The behavior when generating a minion ID. Can specify 'str' or 'func'.
+    # If 'str' is specified (the default), the 'id' option should be set to
+    # the minion ID or left unspecified for default minion ID generation
+    # behavior. If 'func' is specified, the 'id' option should be set to an
+    # exec module function to run to determine the minion ID.
+    'minion_id_type': str,
```

Comment:
I did consider this approach. Given that `master_type` is already implemented, I decided to follow the approach that is already in use (and hence supposedly the safer bet for a design choice in a new feature). If someone else agrees with you to use your approach instead, I am willing to change it.
@cachedout : Do you have an opinion on this?